### ISSUE TITLE
Support Rnw files - i.e.  "knitting" of code chunks before invoking the appropriate latex compiler 

### ISSUE THREAD

**Issue by krishnakumarg1984** [lervag/vimtex#1729]
The feature request is for providing a suitable plug-command as well as overload the current compilation keybinding to support the compilation of `.Rnw` files containing inline code and/or code-chunks

**Describe the solution you'd like**
When editing a `.Rnw` file, the majority of the file (or buffer content) is filled with LaTeX syntax.  A minority component might involve some `inline` computation to dynamically (re)generate results. Compared to the RMarkdown approach, where the majority is markdown, the filetype is primarily LaTeX (RMarkdown also focusses on producing multiple output formats, particularly `html` and `docx`). `knitr` is the analogue of `pythontex` in the R-world, but is much more powerful (can execute commands in ~40 programming languages, including `sed` and `awk` and replace the relevant computational output into the documents).

Hence, `vimtex` should also be the ftplugin for `.Rnw` files so that the motions, syntax, text-objects etc. work as expected.

Furthermore, when the standard `\ll` keybinding is invoked in a project containing `my_computational_results.Rnw`, first vimtex should run `Rscript -e `knitr::knit()` on the current filename which will generate `my_compuational_results.tex`, and then proceed on to invoke `latexmk` as usual.  The entire process should be seamless and automatic for the author of the document.

**Describe alternatives you've considered**
`pythontex`, but really `knitr` is way more powerful. For example, we could just include the highly expressive and superior `ggplot` code for literate programming. Although `pythontex` claims to be extensible to support `R`, knitting seems like a better way to do things for those coming from the R-markdown world. And, it is always nice to have two strong dedicated caterings for the `R` and `python` users.

**Additional context**
`minimal_example.Rnw`

```tex
\documentclass{article}
\usepackage[T1]{fontenc}

\begin{document}

Here is a code chunk.

<<foo, fig.height=4>>=
1 + 1
par(mar = c(4, 4, .2, .2))
plot(rnorm(100))
@

You can also write inline expressions, e.g. $\pi=\Sexpr{pi}$,
and \Sexpr{1.9910214e28} is a big number.

\end{document}
```

**Comment by lervag** [comment]
Yet again, this is interesting. The bundled `ftplugin/rnoweb.vim` already does `runtime! ftplugin/tex.vim` and thus loads vimtex. However, it does not work perfectly, in the sense that it won't e.g. detect multifile projects properly (the functions to find the main file looks for `.tex` files). But for single file projects, this won't be a problem.

I think we should start "light", and I propose the following:

- [ ] Update `vimtex#init()` to detect that it is working on `rnoweb` files and fix multifile project support. This should also make sure that we differentiate the `.Rnw` and `.tex` files in the data blob so that the `.tex` file is available for `latexmk` compilation.
- [ ] Support `rnoweb` compilation. I would actually propose to add this as a separate command. The reason: `\ll` by default starts latexmk with continuous compilation. It works as expected for `.tex` to `.pdf`. No reason to change it. However, we need to add `.Rnw` to `.tex`. Perhaps it is possible to plug this into `latexmk` in some way?

----

Thus, in the first "iteration", we ignore syntax. Let's add that at the second stage. Sounds OK?

It would be helpful if you could give a more detailed "workflow" that _should_ work, but that currently does not work. This would outline which features are not working properly now, in case I have not detected it properly.

**Comment by lervag** [comment]
Did you look into [Nvim-R](https://raw.githubusercontent.com/jalvesaq/Nvim-R/master/doc/Nvim-R.txt)? It seems to have quite a lot of features and support e.g. compilation with latexmk and so on.

**Comment by krishnakumarg1984** [comment]
Yes. I already looked at `nvim-R`. Its a pretty heavy solution and does not neatly fill my current use-case. And it is more suitable for approaching from the other end (my other issue report here in vimtex), wherein the starting format is `Rmarkdown`.

However, `knitr` is very general (a total workhorse really!)

* I am working on a paper with 9 sections, 8 of which contain just static text, which is to be written in the journal's template format
* Only the `results.tex` file contains some computations (less than 10% of the file's content) which are dynamically to be generated with the full power of `knitr`.  The _knitting_ process will replace the dynamic contents with the standard `tex` files which is the only thing my publisher will accept.

So, this is not achieved by `nvim-R`. `nvim-R` is good for **single-file** R-markdown documents. It doesn't have the facilities to process multi-file **bookdown** files. 

This should be no different than supporting `pythontex`. In `pythontex` we just run `python` and replace whatever dynamic computations with the static text from the python result.  In my opinion, using `R` has an advantage that we can use the incredibly powerful, `ggplot2` package for superior *semantic* plots.

**Comment by lervag** [comment]
> However, `knitr` is very general purpose (a total workhorse indeed!).

I'll probably have to learn more about `knitr` at some point!

> Following your request, this is the workflow. ...

Ok, so, let's see if I understand correctly:

* Your project consists of several LaTeX files, most of which are regular `.tex` files, but where one of these files is a `.Rnw` file that produces the `results.tex` file.
* You want vimtex to become aware of `.Rnw` files in a seamless manner so that when we open e.g. `main.tex`, it should "understand" that `results.tex` must be created from `results.Rnw`. (A possible heuristic would be that if `file.Rnw` exists, then always assume that it is the _real_ source file.)
* Note: If we parse the `main.tex` LaTeX structure, it will not really _see_ the `.Rnw` files, because e.g. `\include{file}` will never include a `.Rnw` file. Thus, when we parse the source file structure we need to add a layer of complexity in which we differentiate the _compiled_ files and the _source_ files. Whenever a file is included, we must first check if there is a `.Rnw` file of the same name, and then assume that it is the actual source file.

> As for `vimtex` support, this should be no different than supporting `pythontex`. In `pythontex` package we just run `python` and replace whatever dynamic computations with the static text from the python result. In my opinion, using `R` has an advantage that we can use the incredibly powerful, `ggplot2` package for superior _semantic_ plots.

No, this is quite different. `pythontex` does not work on separate files, but instead on environments within the same `tex` file. To make it work, we need to update `.latexmkrc` to make latexmk aware of it and to run `python` first. However, for `.Rnw` files we need a more complicated workflow. Unless it is possible to somehow make `latexmk` aware of the `.Rnw` files as the source files for `.tex` files.

> The only difference would be in your parser logic? For `pythontex`, you might have been checking whether that package was loaded. For `Rnw` files, you could look at the file extension to deem it as for processing by `knitr`, and/or parse the special delimiter syntax for inline code and block code chunks? Additionally, your `%!TEX-Root` buffer detection might be at play here? This is to determine if the `.Rnw` file is indeed part of a larger project (and not just a separate standalone file).

Again, the difference is larger, because vimtex does not support `pythontex` directly, and instead relies on configuring latexmk properly.

> > Perhaps it is possible to plug this into latexmk in some way?
> 
> By configuring `latexmkrc`? `latexmk` can run arbitrary external commands as custom dependencies right? I have seen configuration examples for processing `asy` files (as well as creation of indexes and glossaries).

Yes, exactly.

> In that case, do you then plan to document `vimtex's` support for `.Rnw` to be restricted to `latexmk` backend? That might disappoint `arara` users, maybe? But it might be a compromise worth having? Major online providers like `overleaf` anyway use `latexmk` as their compilation backend.

I don't know. I don't have a plan, I'm just looking for possible solutions.

----

One possible way to make this work is to use the `$compiling_cmd` configuration variable in latexmk (see [the manual](https://ctan.uib.no/support/latexmk/latexmk.pdf) for details). It allows to run a custom command before latexmk is actually compiling a document. Thus, you could make a simple script that runs `knitr` on all `.Rnw` files before it actually runs the compilation. I'm not 100% sure it works, because I'm not sure if the command is run async'ed or not. But it is worth a shot!

If you _do_ get this to work, then I think this is worth documenting in vimtex as a first step. Perhaps it would also be possible to add a more "full" solution within vimtex as well.

Also, another thing that seems "sensible" is for vimtex to load "as normal" for `.Rnw` files. It seems we could just load everything as it is, except that the current file name must be set to the real file name.

----

Other references I've found that seems "relevant":

* https://stackoverflow.com/questions/28669967/using-vim-latex-suite-knitr/35389662
  (It seems one could also use something like `entr` for running `knitr` whenever the files change in a similar manner as suggested in the answer in this link.)

**Comment by smilesun** [comment]
how do I activate the Rnw file type for vimtex? For me, vimtex is not activated when I open a Rnw file

**Comment by lervag** [comment]
First: Why do you ask that question _here_? Please open a new issue if you have questions, instead of "bumping" old threads (more than three years old here).

Now, VimTeX is _not_ a plugin for Rnw files. I know some people do load VimTeX, and I guess it does work, but I don't want to personally support that as I don't use Rnw files myself. I think a better place to ask this question would be on Reddit (r/vim or r/neovim) or on vi.stachexchange.com.
